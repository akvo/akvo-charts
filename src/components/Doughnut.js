import React, { useMemo } from 'react';
import { useECharts } from '../hooks';
import styles from '../styles.module.css';

const MAX = 70;

const getOptions = ({ data = [], radius }) => ({
  series: [
    {
      radius,
      data: data.map((item) => ({ name: item.label, value: item.value }))
    }
  ]
});

const Doughnut = ({ config, data, size = 40 }) => {
  const torus = useMemo(() => {
    if (size >= 70) {
      return 0;
    }
    return MAX - size;
  }, [size]);

  const chartRef = useECharts({
    config,
    data,
    getOptions: ({ data }) =>
      getOptions({ data, radius: [`${torus}%`, `${MAX}%`] })
  });

  return (
    <div
      ref={chartRef}
      role="figure"
      className={styles.container}
    />
  );
};

export default Doughnut;
