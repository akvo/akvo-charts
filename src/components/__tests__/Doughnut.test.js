import React from 'react';
import { render, screen } from '@testing-library/react';
import Doughnut from '../Doughnut';
import renderer from 'react-test-renderer';

test('renders Doughnut component', () => {
  const data = [
    { label: 'Category A', value: 30 },
    { label: 'Category B', value: 70 }
  ];

  const config = {
    title: 'Doughnut Chart Example'
  };

  render(
    <Doughnut
      config={config}
      data={data}
      size={50}
    />
  );

  const chartContainer = screen.getByRole('figure');
  expect(chartContainer).toBeInTheDocument();
});

test('matches Doughnut snapshot', () => {
  const data = [
    { label: 'Category A', value: 30 },
    { label: 'Category B', value: 70 }
  ];

  const config = {
    title: 'Doughnut Chart Example'
  };

  const tree = renderer
    .create(
      <Doughnut
        config={config}
        data={data}
        size={50}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
