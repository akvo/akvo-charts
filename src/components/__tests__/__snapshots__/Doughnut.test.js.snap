// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches Doughnut snapshot 1`] = `
<div
  className="container"
  role="figure"
/>
`;
